lib/Runtime/Library/JsBuiltInEngineInterfaceExtensionObject.cpp

```diff
@@ -299,46 +265,46 @@ namespace Js
     {
         EngineInterfaceObject_CommonFunctionProlog(function, callInfo);
 
-        AssertOrFailFast(args.Info.Count == 3 && TaggedInt::Is(args.Values[1]) && VarIs<ScriptFunction>(args.Values[2]));
+        AssertOrFailFast(args.Info.Count == 3 && VarIs<ScriptFunction>(args.Values[2]));
 
         JavascriptLibrary * library = scriptContext->GetLibrary();
 
-        FunctionKind funcKind = static_cast<FunctionKind>(TaggedInt::ToInt32(args.Values[1]));
-        AssertOrFailFast(funcKind >= (FunctionKind)0 && funcKind < FunctionKind::Max);
+        // process function name and get property ID
+        JavascriptString* methodString = VarTo<JavascriptString>(args.Values[1]);
+        PropertyRecord* methodPropertyRecord = nullptr;
+        methodString->GetPropertyRecord((PropertyRecord const **) &methodPropertyRecord, false);
+        PropertyId methodPropID = methodPropertyRecord->GetPropertyId();
 
         DynamicObject *installTarget = nullptr;
         bool isStatic = false;
-        PropertyId methodPropID = PropertyIds::_none;
-        PropertyString *methodPropString = nullptr;
         PropertyString *classPropString = nullptr;
         JavascriptString *fullName = nullptr;
         JavascriptString *dot = library->GetDotString();
-        switch (funcKind)
+        JsBuiltInFile current = (static_cast<JsBuiltInEngineInterfaceExtensionObject*>(scriptContext->GetLibrary()
+            ->GetEngineInterfaceObject()->GetEngineExtension(EngineInterfaceExtensionKind_JsBuiltIn)))->current;
+
+        // determine what object this function is being added too
+        switch (current)
         {
-#define VALUE(ClassName, methodName, propertyType) \
-        case FunctionKind::ClassName##_##methodName: \
-            isStatic = static_cast<bool>(IsPropertyTypeStatic::##propertyType); \
-            installTarget = library->Get##ClassName##propertyType##(); \
-            methodPropID = PropertyIds::methodName; \
-            methodPropString = scriptContext->GetPropertyString(methodPropID); \
-            classPropString = scriptContext->GetPropertyString(PropertyIds::ClassName); \
-            break;
-FUNCTIONKIND_VALUES(VALUE)
-#undef VALUE
-        default:
-            AssertOrFailFastMsg(false, "funcKind should never be outside the range of projected values");
+            #define file(class, type, obj) \
+            case class##_##type: \
+                isStatic = static_cast<bool>(IsTypeStatic::##type); \
+                installTarget = library->Get##obj##(); \
+                classPropString = scriptContext->GetPropertyString(PropertyIds::class); \
+                break;
+            JsBuiltIns(file)
```

Comment:
Previous switch relied on the FunctionKind object to determine both the name of the function AND the object it should be set as a method on.

Now the object is determined here based on the `current` file being loaded and the function name is found above from the string passed in.